src/devices_models/devices/common/cost_functions.jl

```diff
@@ -49,6 +49,17 @@ function AddCostSpec(
     error("AddCostSpec is not implemented for $T / $U")
 end
 
+function add_service_variables!(
+    spec::AddCostSpec,
+    services::Vector{T},
```

Comment:
Given that the function is called add_service_variables you may not need T for dispatch.  Making services untyped would allow easier use of the function because users could pass any iterable.